addons/bus/models/bus.py

```diff
@@ -106,93 +105,61 @@ def _poll(self, channels, last=0, options=None):
-        channels = [channel_with_db(dbname, channel) for channel in channels]
-        if timeout is None:
-            timeout = TIMEOUT
-        if options is None:
-            options = {}
-        # Dont hang ctrl-c for a poll request, we need to bypass private
-        # attribute access because we dont know before starting the thread that
-        # it will handle a longpolling request
-        if not odoo.evented:
-            current = threading.current_thread()
-            current._daemonic = True
-            # rename the thread to avoid tests waiting for a longpolling
-            current.name = f"openerp.longpolling.request.{current.ident}"
-
-        registry = odoo.registry(dbname)
-
-        # immediatly returns if past notifications exist
-        with registry.cursor() as cr:
-            env = api.Environment(cr, SUPERUSER_ID, {})
-            notifications = env['bus.bus']._poll(channels, last, options)
 
-        # immediatly returns in peek mode
-        if options.get('peek'):
-            return dict(notifications=notifications, channels=channels)
+class BusSubscription:
+    def __init__(self, dbname, channels, last):
+        self.last_notification_id = last
+        self.channels = channels
+        self.dbname = dbname
 
-        # or wait for future ones
-        if not notifications:
-            if not self.started:
-                # Lazy start of events listener
-                self.start()
 
-            event = self.Event()
-            for channel in channels:
-                self.channels.setdefault(hashable(channel), set()).add(event)
-            try:
-                event.wait(timeout=timeout)
-                with registry.cursor() as cr:
-                    env = api.Environment(cr, SUPERUSER_ID, {})
-                    notifications = env['bus.bus']._poll(channels, last, options)
-            except Exception:
-                # timeout
-                pass
-            finally:
-                # gc pointers to event
-                for channel in channels:
-                    channel_events = self.channels.get(hashable(channel))
-                    if channel_events and event in channel_events:
-                        channel_events.remove(event)
-        return notifications
+class ImDispatch(threading.Thread):
+    def __init__(self):
+        super().__init__()
+        self._daemonic = True
+        self.name = f'{__name__}.Bus'
+        self._ws_to_subscription = {}
+
+    def _dispatch_notifications(self, websocket, subscription):
+        with odoo.registry(subscription.dbname).cursor() as cr:
+            env = api.Environment(cr, SUPERUSER_ID, {})
+            notifications = env['bus.bus']._poll(
+                subscription.channels, subscription.last_notification_id)
+            if notifications:
+                subscription.last_notification_id = notifications[-1]['id']
+                websocket.send(notifications)
+
+    def subscribe(self, channels):
```

Comment:
I don't like this function uses the global ws_request object, from the outside you read `dispatch.subscribre(channels)` and you don't understand that the channels are registered for the current websocket. IMO the controller should give the current websocket as argument and this file should not import it.